core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1006,6 +1005,16 @@ private void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, Ex
 
             // Update log start offset with the computed value after retention cleanup is done
             remoteLogRetentionHandler.logStartOffset.ifPresent(offset -> handleLogStartOffsetUpdate(topicIdPartition.topicPartition(), offset));
+
+            List<String> undeletedSegments = new ArrayList<>();
+            for (RemoteLogSegmentMetadata segmentMetadata : segmentsToDelete) {
+                if (!remoteLogRetentionHandler.deleteRemoteLogSegment(segmentMetadata, x -> !isCancelled() && isLeader())) {
+                    undeletedSegments.add(segmentMetadata.remoteLogSegmentId().toString());
+                }
+            }
+            if (!undeletedSegments.isEmpty()) {
+                logger.error("The following remote segments could not be deleted: {}", String.join(",", undeletedSegments));
```

Comment:
not an error since it's recoverable and expected when leadership has transitioned over. Perhaps a info.